src/btree/bt_sync.c

```diff
@@ -377,18 +422,32 @@ __wt_sync_file(WT_SESSION_IMPL *session, WT_CACHE_OP syncop)
         if (is_las) {
             LF_CLR(WT_READ_CACHE);
             LF_SET(WT_READ_CACHE_LEAF);
+            memset(&ref_list, 0, sizeof(ref_list));
         }
 
         for (;;) {
             WT_ERR(__sync_dup_walk(session, walk, flags, &prev));
             WT_ERR(__wt_tree_walk(session, &walk, flags));
 
-            if (walk == NULL)
+            if (walk == NULL) {
+                if (is_las) {
+                    for (i = 0; i < ref_list.entry; i++) {
+                        WT_IGNORE_RET_BOOL(__wt_page_evict_urgent(session, ref_list.list[i]));
+                        WT_ERR(__wt_page_release(session, ref_list.list[i], flags));
+                        WT_STAT_CONN_INCR(session, hs_gc_pages_evict);
+                        __wt_verbose(session, WT_VERB_CHECKPOINT_GC,
+                          "%p: is an in-memory obsolete page, added to urgent eviction queue.",
+                          (void *)ref_list.list[i]);
+                    }
+                    __sync_ref_list_free(session, &ref_list);
+                }
```

Comment:
If we're going to be storing a list of pages to be evicted, could we get rid of that weird stutter-step thing we're doing with `__sync_dup_walk()`? That is, just walk through the tree normally, storing `WT_REFs` for later deletion, and that "duplicate the walk" magic goes away. (That code was remarkably difficult to get right, and I'd love to see it gone.)

Reply:
The logic of storing is only for the history store, this may not work for the general-purpose logic of `__sync_dup_walk()` current usage?

Reply:
We might overwhelm the urgent queue (I'm not sure how that works so it may or may not be an issue). I'd love to get rid of that stutter-step walk, though.